### PR TITLE
Allow runner to check service connection in background.

### PR DESCRIPTION
**Pull request by TingluoHuang**
The server can request runner to do connectivity check as need to debug network issue between the runner and the service.

### CODE REVIEW

File: src/Sdk/DTWebApi/WebApi/ServiceConnectivityCheck.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.Serialization;
+using Newtonsoft.Json;
+
+namespace GitHub.DistributedTask.WebApi
+{
+    [DataContract]
+    public class ServiceConnectivityCheckInput
```
Comment (TingluoHuang) [inline src/Sdk/DTWebApi/WebApi/ServiceConnectivityCheck.cs:9]: i will copy the contract to the service side.

File: src/Runner.Worker/JobExtension.cs [hunk 2]
```diff
@@ -781,6 +828,91 @@ private async Task CheckDiskSpaceAsync(IExecutionContext context, CancellationTo
...
         }
 
+        private async Task CheckServiceConnectivityAsync(IExecutionContext context, CancellationToken token)
+        {
+            var foo = new ServiceConnectivityCheckInput() { IntervalInSecond = 30, RequestTimeoutInSecond = 30 };
+            foo.Endpoints.Add("pipelines", "https://pipelinesghubeus20.actions.githubusercontent.com/_apis/connectiondata");
+            foo.Endpoints.Add("broker", "https://broker.actions.githubusercontent.com/health");
+            foo.Endpoints.Add("results", "https://results-receiver.actions.githubusercontent.com/health");
+            var fooJson = StringUtil.ConvertToJson(foo);
+            context.Global.Variables.Set(WellKnownDistributedTaskVariables.RunnerServiceConnectivityTest, fooJson);
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:838]: this is an example input, i will remove this before merge the PR.

File: src/Runner.Worker/JobExtension.cs [hunk 3]
```diff
@@ -717,33 +739,58 @@ public async Task FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRe
             }
         }
 
-        private async Task<string> CheckConnectivity(string endpointUrl)
+        private async Task<CheckResult> CheckConnectivity(string endpointUrl, string accessToken, int timeoutInSeconds, CancellationToken token)
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:742]: I make this more generic so i can reuse between the current 1 time connection check and the backend long-running connection check.

File: src/Runner.Worker/JobExtension.cs [hunk 4]
```diff
@@ -717,33 +739,58 @@ public async Task FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRe
...
         {
             Trace.Info($"Check server connectivity for {endpointUrl}.");
-            string result = string.Empty;
-            using (var timeoutTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(5)))
+            CheckResult result = new CheckResult() { EndpointUrl = endpointUrl };
+            var stopwatch = Stopwatch.StartNew();
+            using (var timeoutTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(timeoutInSeconds)))
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:747]: the existing use case will pass `5` as method input.

File: src/Runner.Worker/JobExtension.cs [hunk 5]
```diff
@@ -717,33 +739,58 @@ public async Task FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRe
...
+            using (var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token, timeoutTokenSource.Token))
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:748]: the existing use code will pass `CancellationToken.None`

File: src/Runner.Worker/JobExtension.cs [hunk 6]
```diff
@@ -717,33 +739,58 @@ public async Task FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRe
...
             {
                 try
                 {
                     using (var httpClientHandler = HostContext.CreateHttpClientHandler())
                     using (var httpClient = new HttpClient(httpClientHandler))
                     {
                         httpClient.DefaultRequestHeaders.UserAgent.AddRange(HostContext.UserAgents);
-                        var response = await httpClient.GetAsync(endpointUrl, timeoutTokenSource.Token);
-                        result = $"{endpointUrl}: {response.StatusCode}";
+                        if (!string.IsNullOrEmpty(accessToken))
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:756]: existing code will pass empty string.

File: src/Runner.Worker/JobExtension.cs [hunk 7]
```diff
@@ -717,33 +739,58 @@ public async Task FinalizeJob(IExecutionContext jobContext, Pipelines.AgentJobRe
...
                 }
+                catch (Exception ex) when (ex is OperationCanceledException && token.IsCancellationRequested)
+                {
+                    Trace.Error($"Request canceled during connectivity check: {ex}");
+                    result.StatusCode = "canceled";
+                }
                 catch (Exception ex) when (ex is OperationCanceledException && timeoutTokenSource.IsCancellationRequested)
                 {
                     Trace.Error($"Request timeout during connectivity check: {ex}");
-                    result = $"{endpointUrl}: timeout";
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:739]: we will reconstruct the same result from the caller.

File: src/Runner.Worker/JobExtension.cs [hunk 8]
```diff
@@ -781,6 +828,91 @@ private async Task CheckDiskSpaceAsync(IExecutionContext context, CancellationTo
...
+        {
+            var foo = new ServiceConnectivityCheckInput() { IntervalInSecond = 30, RequestTimeoutInSecond = 30 };
+            foo.Endpoints.Add("pipelines", "https://pipelinesghubeus20.actions.githubusercontent.com/_apis/connectiondata");
+            foo.Endpoints.Add("broker", "https://broker.actions.githubusercontent.com/health");
+            foo.Endpoints.Add("results", "https://results-receiver.actions.githubusercontent.com/health");
+            var fooJson = StringUtil.ConvertToJson(foo);
+            context.Global.Variables.Set(WellKnownDistributedTaskVariables.RunnerServiceConnectivityTest, fooJson);
+
+            var connectionTest = context.Global.Variables.Get(WellKnownDistributedTaskVariables.RunnerServiceConnectivityTest);
+            if (string.IsNullOrEmpty(connectionTest))
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:841]: this is the featureflag to control the test.

File: src/Runner.Worker/JobExtension.cs [hunk 9]
```diff
@@ -781,6 +828,91 @@ private async Task CheckDiskSpaceAsync(IExecutionContext context, CancellationTo
...
+                {
+                    await Task.Delay(TimeSpan.FromSeconds(checkConnectivityInfo.IntervalInSecond), token);
+                }
+                catch (TaskCanceledException)
+                {
+                    // ignore
+                }
+            }
+
+            var telemetryData = StringUtil.ConvertToJson(testResult, Formatting.None);
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:911]: don't add indent to save space.

**Review by rentziass — APPROVED** [review]

**Review by rentziass — APPROVED** [review]

**Pull request merged**